diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
index 6ecca9cb98a..e397dc0e113 100644
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -4841,20 +4841,35 @@ NTSTATUS WINAPI NtReadVirtualMemory( HANDLE process, const void *addr, void *buf
 
     if (virtual_check_buffer_for_write( buffer, size ))
     {
-        SERVER_START_REQ( read_process_memory )
+        if (GetCurrentProcess() == process)
         {
-            req->handle = wine_server_obj_handle( process );
-            req->addr   = wine_server_client_ptr( addr );
-            wine_server_set_reply( req, buffer, size );
-            if ((status = wine_server_call( req ))) size = 0;
+            if (!virtual_check_buffer_for_read(addr, size))
+            {
+                status = STATUS_ACCESS_VIOLATION;
+                size = 0;
+            }
+
+            size = virtual_uninterrupted_read_memory(addr, buffer, size);
+            status = STATUS_SUCCESS;
+        }
+        else
+        {
+            SERVER_START_REQ( read_process_memory )
+            {
+                req->handle = wine_server_obj_handle( process );
+                req->addr   = wine_server_client_ptr( addr );
+                wine_server_set_reply( req, buffer, size );
+                if ((status = wine_server_call( req ))) size = 0;
+            }
+            SERVER_END_REQ;
         }
-        SERVER_END_REQ;
     }
     else
     {
         status = STATUS_ACCESS_VIOLATION;
         size = 0;
     }
+
     if (bytes_read) *bytes_read = size;
     return status;
 }
